diff --git a/CMakeLists.txt b/CMakeLists.txt
index 91e47a0..4aa2bdf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,10 +11,11 @@
 ###########################################################
 
 cmake_minimum_required(VERSION 3.12.4)
-project(direct_lidar_inertial_odometry)
+project(direct_lidar_inertial_odometry VERSION 1.1.1)
 
 add_compile_options(-std=c++17)
 set(CMAKE_BUILD_TYPE "Release")
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
 find_package( PCL REQUIRED )
 include_directories(${PCL_INCLUDE_DIRS})
@@ -33,50 +34,7 @@ else(OPENMP_FOUND)
   message("ERROR: OpenMP could not be found.")
 endif(OPENMP_FOUND)
 
-set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
-set(THREADS_PREFER_PTHREAD_FLAG TRUE)
-find_package(Threads REQUIRED)
-
-find_package(catkin REQUIRED COMPONENTS
-  roscpp
-  std_msgs
-  sensor_msgs
-  geometry_msgs
-  nav_msgs
-  pcl_ros
-  message_generation
-)
-
-add_service_files(
-  DIRECTORY srv
-  FILES
-  save_pcd.srv
-)
-
-generate_messages(
-  DEPENDENCIES
-  sensor_msgs
-  std_msgs
-  geometry_msgs
-)
-
-catkin_package(
-  CATKIN_DEPENDS
-    roscpp
-    std_msgs
-    sensor_msgs
-    geometry_msgs
-    pcl_ros
-  INCLUDE_DIRS
-    include
-  LIBRARIES
-    ${PROJECT_NAME}
-    nano_gicp
-    nanoflann
-)
-
 include_directories(include)
-include_directories(SYSTEM ${catkin_INCLUDE_DIRS})
 
 # Not all machines have <cpuid.h> available
 set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
@@ -91,29 +49,24 @@ endif()
 add_library(nanoflann STATIC
   src/nano_gicp/nanoflann.cc
 )
-target_link_libraries(nanoflann ${PCL_LIBRARIES})
+target_link_libraries(nanoflann PUBLIC ${PCL_LIBRARIES})
+target_include_directories(nanoflann PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
 
 # NanoGICP
 add_library(nano_gicp STATIC
   src/nano_gicp/lsq_registration.cc
   src/nano_gicp/nano_gicp.cc
 )
-target_link_libraries(nano_gicp ${PCL_LIBRARIES} ${OpenMP_LIBS} nanoflann)
+target_link_libraries(nano_gicp PUBLIC ${PCL_LIBRARIES} ${OpenMP_LIBS} nanoflann)
+target_include_directories(nano_gicp PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
 
 # Odometry Node
-add_executable(dlio_odom_node src/dlio/odom_node.cc src/dlio/odom.cc)
-add_dependencies(dlio_odom_node ${catkin_EXPORTED_TARGETS} ${PROJECT_NAME}_generate_messages_cpp)
+add_library(dlio_odom_node src/dlio/odom.cc)
 target_compile_options(dlio_odom_node PRIVATE ${OpenMP_FLAGS})
-target_link_libraries(dlio_odom_node ${catkin_LIBRARIES} ${PCL_LIBRARIES} ${OpenMP_LIBS} Threads::Threads nano_gicp)
+target_link_libraries(dlio_odom_node PUBLIC ${PCL_LIBRARIES} ${OpenMP_LIBS} nano_gicp)
+target_include_directories(dlio_odom_node PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
 
 # Mapping Node
-add_executable (dlio_map_node src/dlio/map_node.cc src/dlio/map.cc)
-add_dependencies(dlio_map_node ${catkin_EXPORTED_TARGETS} ${PROJECT_NAME}_generate_messages_cpp)
-target_compile_options(dlio_map_node PRIVATE ${OpenMP_FLAGS})
-target_link_libraries(dlio_map_node ${catkin_LIBRARIES} ${PCL_LIBRARIES} ${OpenMP_LIBS} Threads::Threads)
-
-# Binaries
-install( TARGETS dlio_odom_node dlio_map_node
-         DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION} )
-install( DIRECTORY cfg launch
-         DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION} )
+# add_executable (dlio_map_node src/dlio/map_node.cc src/dlio/map.cc)
+# target_compile_options(dlio_map_node PRIVATE ${OpenMP_FLAGS})
+# target_link_libraries(dlio_map_node ${PCL_LIBRARIES} ${OpenMP_LIBS} Threads::Threads)
diff --git a/include/dlio/dlio.h b/include/dlio/dlio.h
index 47386ff..0bda740 100644
--- a/include/dlio/dlio.h
+++ b/include/dlio/dlio.h
@@ -44,16 +44,6 @@ std::string to_string_with_precision(const T a_value, const int n = 6)
     return out.str();
 }
 
-// ROS
-#include <ros/ros.h>
-#include <nav_msgs/Odometry.h>
-#include <geometry_msgs/PoseStamped.h>
-#include <geometry_msgs/PoseArray.h>
-#include <nav_msgs/Path.h>
-#include <sensor_msgs/Imu.h>
-#include <sensor_msgs/PointCloud2.h>
-#include <tf2_ros/transform_broadcaster.h>
-
 // BOOST
 #include <boost/format.hpp>
 #include <boost/circular_buffer.hpp>
@@ -69,14 +59,9 @@ std::string to_string_with_precision(const T a_value, const int n = 6)
 #include <pcl/io/pcd_io.h>
 #include <pcl/surface/concave_hull.h>
 #include <pcl/surface/convex_hull.h>
-#include <pcl_conversions/pcl_conversions.h>
-#include <pcl_ros/impl/transforms.hpp>
-#include <pcl_ros/point_cloud.h>
-#include <pcl_ros/transforms.h>
 
 // DLIO
 #include <nano_gicp/nano_gicp.h>
-#include <direct_lidar_inertial_odometry/save_pcd.h>
 
 namespace dlio {
   enum class SensorType { OUSTER, VELODYNE, HESAI, LIVOX, UNKNOWN };
@@ -84,7 +69,7 @@ namespace dlio {
   class OdomNode;
   class MapNode;
 
-  struct Point {
+  struct EIGEN_ALIGN16 Point {
     Point(): data{0.f, 0.f, 0.f, 1.f} {}
 
     PCL_ADD_POINT4D;
@@ -96,7 +81,7 @@ namespace dlio {
                        // (Livox) absolute timestamp in (seconds * 10e9)
     };
     EIGEN_MAKE_ALIGNED_OPERATOR_NEW
-  } EIGEN_ALIGN16;
+  };
 }
 
 POINT_CLOUD_REGISTER_POINT_STRUCT(dlio::Point,
diff --git a/include/dlio/odom.h b/include/dlio/odom.h
index c79ebd7..2c91f67 100644
--- a/include/dlio/odom.h
+++ b/include/dlio/odom.h
@@ -10,35 +10,111 @@
  *                                                         *
  ***********************************************************/
 
+#include <cmath>
+
 #include "dlio/dlio.h"
 
 class dlio::OdomNode {
 
 public:
+  struct Params {
+    bool verbose = false;
+
+    bool deskew = true;
+    double gravity = 9.80665;
+    bool time_offset = false;
+
+    double keyframe_thresh_dist = 0.1;
+    double keyframe_thresh_rot = 1.0;
+
+    int submap_knn = 10;
+    int submap_kcv = 10;
+    int submap_kcc = 10;
+
+    bool densemap_filtered = true;
+    bool wait_until_move = false;
+
+    double crop_size = 1.0;
+
+    bool vf_use = true;
+    double vf_res = 0.05;
+
+    bool adaptive_params = true;
+
+    Eigen::Vector3f extrinsics_baselink2imu_t = Eigen::Vector3f::Zero();
+    Eigen::Matrix3f extrinsics_baselink2imu_R = Eigen::Matrix3f::Identity();
+    Eigen::Vector3f extrinsics_baselink2lidar_t = Eigen::Vector3f::Zero();
+    Eigen::Matrix3f extrinsics_baselink2lidar_R = Eigen::Matrix3f::Identity();
+
+    bool calibrate_gyro = true;
+    bool calibrate_accel = true;
+    double imu_calib_time = 3.0;
+    int imu_buffer_size = 2000;
+
+    bool gravity_align = true;
+    bool imu_calibrate = true;
+
+    int gicp_min_num_points = 100;
+    int gicp_k_correspondences = 20;
+    double gicp_max_corr_dist = std::sqrt(std::numeric_limits<double>::max());
+    int gicp_max_iter = 64;
+    double gicp_transformation_ep = 0.0005;
+    double gicp_rotation_ep = 0.0005;
+    double gicp_init_lambda_factor = 1e-9;
+
+    double geo_Kp = 1.0;
+    double geo_Kv = 1.0;
+    double geo_Kq = 1.0;
+    double geo_Kab = 1.0;
+    double geo_Kgb = 1.0;
+    double geo_abias_max = 1.0;
+    double geo_gbias_max = 1.0;
+  };
 
-  OdomNode(ros::NodeHandle node_handle);
-  ~OdomNode();
+  struct ImuMeas {
+    double stamp;
+    double dt; // defined as the difference between the current and the previous
+               // measurement
+    Eigen::Vector3f ang_vel;
+    Eigen::Vector3f lin_accel;
+  };
 
-  void start();
+  struct ImuBias {
+    Eigen::Vector3f gyro;
+    Eigen::Vector3f accel;
+  };
 
-private:
+  struct Frames {
+    Eigen::Vector3f b;
+    Eigen::Vector3f w;
+  };
 
-  struct State;
-  struct ImuMeas;
+  struct Velocity {
+    Frames lin;
+    Frames ang;
+  };
 
-  void getParams();
+  struct State {
+    Eigen::Vector3f p;    // position in world frame
+    Eigen::Quaternionf q; // orientation in world frame
+    Velocity v;
+    ImuBias b; // imu biases in body frame
+  };
 
-  void callbackPointCloud(const sensor_msgs::PointCloud2ConstPtr& pc);
-  void callbackImu(const sensor_msgs::Imu::ConstPtr& imu);
+  OdomNode(const Params &params);
+  ~OdomNode();
 
-  void publishPose(const ros::TimerEvent& e);
+  void start();
+  State getState() const { return state; };
+
+  void callbackPointCloud(const pcl::PointCloud<PointType>::ConstPtr &pc,
+                          double stamp);
+  void callbackImu(const ImuMeas &imu);
 
-  void publishToROS(pcl::PointCloud<PointType>::ConstPtr published_cloud, Eigen::Matrix4f T_cloud);
-  void publishCloud(pcl::PointCloud<PointType>::ConstPtr published_cloud, Eigen::Matrix4f T_cloud);
-  void publishKeyframe(std::pair<std::pair<Eigen::Vector3f, Eigen::Quaternionf>,
-                       pcl::PointCloud<PointType>::ConstPtr> kf, ros::Time timestamp);
+private:
+  void getParams(const Params &params);
 
-  void getScanFromROS(const sensor_msgs::PointCloud2ConstPtr& pc);
+  void getScanFromROS(const pcl::PointCloud<PointType>::ConstPtr &pc, double stamp);
   void preprocessPoints();
   void deskewPointcloud();
   void initializeInputTarget();
@@ -70,7 +146,7 @@ private:
   void computeSpaciousness();
   void computeDensity();
 
-  sensor_msgs::Imu::Ptr transformImu(const sensor_msgs::Imu::ConstPtr& imu);
+  ImuMeas transformImu(const ImuMeas &imu);
 
   void updateKeyframes();
   void computeConvexHull();
@@ -82,27 +158,6 @@ private:
 
   void debug();
 
-  ros::NodeHandle nh;
-  ros::Timer publish_timer;
-
-  // Subscribers
-  ros::Subscriber lidar_sub;
-  ros::Subscriber imu_sub;
-
-  // Publishers
-  ros::Publisher odom_pub;
-  ros::Publisher pose_pub;
-  ros::Publisher path_pub;
-  ros::Publisher kf_pose_pub;
-  ros::Publisher kf_cloud_pub;
-  ros::Publisher deskewed_pub;
-
-  // ROS Msgs
-  nav_msgs::Odometry odom_ros;
-  geometry_msgs::PoseStamped pose_ros;
-  nav_msgs::Path path_ros;
-  geometry_msgs::PoseArray kf_pose_ros;
-
   // Flags
   std::atomic<bool> dlio_initialized;
   std::atomic<bool> first_valid_scan;
@@ -113,12 +168,6 @@ private:
   std::atomic<bool> deskew_status;
   std::atomic<int> deskew_size;
 
-  // Threads
-  std::thread publish_thread;
-  std::thread publish_keyframe_thread;
-  std::thread metrics_thread;
-  std::thread debug_thread;
-
   // Trajectory
   std::vector<std::pair<Eigen::Vector3f, Eigen::Quaternionf>> trajectory;
   double length_traversed;
@@ -126,7 +175,7 @@ private:
   // Keyframes
   std::vector<std::pair<std::pair<Eigen::Vector3f, Eigen::Quaternionf>,
                         pcl::PointCloud<PointType>::ConstPtr>> keyframes;
-  std::vector<ros::Time> keyframe_timestamps;
+  std::vector<double> keyframe_timestamps;
   std::vector<std::shared_ptr<const nano_gicp::CovarianceList>> keyframe_normals;
   std::vector<Eigen::Matrix4f, Eigen::aligned_allocator<Eigen::Matrix4f>> keyframe_transformations;
   std::mutex keyframes_mutex;
@@ -134,12 +183,6 @@ private:
   // Sensor Type
   dlio::SensorType sensor;
 
-  // Frames
-  std::string odom_frame;
-  std::string baselink_frame;
-  std::string lidar_frame;
-  std::string imu_frame;
-
   // Preprocessing
   pcl::CropBox<PointType> crop;
   pcl::VoxelGrid<PointType> voxel;
@@ -173,7 +216,7 @@ private:
   std::mutex main_loop_running_mutex;
 
   // Timestamps
-  ros::Time scan_header_stamp;
+  double scan_header_stamp;
   double scan_stamp;
   double prev_scan_stamp;
   double scan_dt;
@@ -206,17 +249,12 @@ private:
   }; Extrinsics extrinsics;
 
   // IMU
-  ros::Time imu_stamp;
+  double imu_stamp;
   double first_imu_stamp;
   double prev_imu_stamp;
   double imu_dp, imu_dq_deg;
 
-  struct ImuMeas {
-    double stamp;
-    double dt; // defined as the difference between the current and the previous measurement
-    Eigen::Vector3f ang_vel;
-    Eigen::Vector3f lin_accel;
-  }; ImuMeas imu_meas;
+  ImuMeas imu_meas;
 
   boost::circular_buffer<ImuMeas> imu_buffer;
   std::mutex mtx_imu;
@@ -238,27 +276,7 @@ private:
   }; Geo geo;
 
   // State Vector
-  struct ImuBias {
-    Eigen::Vector3f gyro;
-    Eigen::Vector3f accel;
-  };
-
-  struct Frames {
-    Eigen::Vector3f b;
-    Eigen::Vector3f w;
-  };
-
-  struct Velocity {
-    Frames lin;
-    Frames ang;
-  };
-
-  struct State {
-    Eigen::Vector3f p; // position in world frame
-    Eigen::Quaternionf q; // orientation in world frame
-    Velocity v;
-    ImuBias b; // imu biases in body frame
-  }; State state;
+  State state;
 
   struct Pose {
     Eigen::Vector3f p; // position in world frame
@@ -279,7 +297,7 @@ private:
   int numProcessors;
 
   // Parameters
-  std::string version_;
+  std::string version_ = "1.1.1";
   int num_threads_;
   bool verbose;
 
diff --git a/src/dlio/odom.cc b/src/dlio/odom.cc
index bceae37..c0c625f 100644
--- a/src/dlio/odom.cc
+++ b/src/dlio/odom.cc
@@ -12,9 +12,9 @@
 
 #include "dlio/odom.h"
 
-dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
+dlio::OdomNode::OdomNode(const Params &params) {
 
-  this->getParams();
+  this->getParams(params);
 
   this->num_threads_ = omp_get_max_threads();
 
@@ -26,20 +26,6 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
   this->deskew_status = false;
   this->deskew_size = 0;
 
-  this->lidar_sub = this->nh.subscribe("pointcloud", 1,
-      &dlio::OdomNode::callbackPointCloud, this, ros::TransportHints().tcpNoDelay());
-  this->imu_sub = this->nh.subscribe("imu", 1000,
-      &dlio::OdomNode::callbackImu, this, ros::TransportHints().tcpNoDelay());
-
-  this->odom_pub     = this->nh.advertise<nav_msgs::Odometry>("odom", 1, true);
-  this->pose_pub     = this->nh.advertise<geometry_msgs::PoseStamped>("pose", 1, true);
-  this->path_pub     = this->nh.advertise<nav_msgs::Path>("path", 1, true);
-  this->kf_pose_pub  = this->nh.advertise<geometry_msgs::PoseArray>("kf_pose", 1, true);
-  this->kf_cloud_pub = this->nh.advertise<sensor_msgs::PointCloud2>("kf_cloud", 1, true);
-  this->deskewed_pub = this->nh.advertise<sensor_msgs::PointCloud2>("deskewed", 1, true);
-
-  this->publish_timer = this->nh.createTimer(ros::Duration(0.01), &dlio::OdomNode::publishPose, this);
-
   this->T = Eigen::Matrix4f::Identity();
   this->T_prior = Eigen::Matrix4f::Identity();
   this->T_corr = Eigen::Matrix4f::Identity();
@@ -67,10 +53,10 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
   this->first_imu_stamp = 0.;
   this->prev_imu_stamp = 0.;
 
-  this->original_scan = pcl::PointCloud<PointType>::ConstPtr (boost::make_shared<const pcl::PointCloud<PointType>>());
-  this->deskewed_scan = pcl::PointCloud<PointType>::ConstPtr (boost::make_shared<const pcl::PointCloud<PointType>>());
-  this->current_scan = pcl::PointCloud<PointType>::ConstPtr (boost::make_shared<const pcl::PointCloud<PointType>>());
-  this->submap_cloud = pcl::PointCloud<PointType>::ConstPtr (boost::make_shared<const pcl::PointCloud<PointType>>());
+  this->original_scan = pcl::PointCloud<PointType>::ConstPtr (std::make_shared<const pcl::PointCloud<PointType>>());
+  this->deskewed_scan = pcl::PointCloud<PointType>::ConstPtr (std::make_shared<const pcl::PointCloud<PointType>>());
+  this->current_scan = pcl::PointCloud<PointType>::ConstPtr (std::make_shared<const pcl::PointCloud<PointType>>());
+  this->submap_cloud = pcl::PointCloud<PointType>::ConstPtr (std::make_shared<const pcl::PointCloud<PointType>>());
 
   this->num_processed_keyframes = 0;
 
@@ -79,7 +65,6 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
 
   this->first_scan_stamp = 0.;
   this->elapsed_time = 0.;
-  this->length_traversed;
 
   this->convex_hull.setDimension(3);
   this->concave_hull.setDimension(3);
@@ -157,151 +142,81 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
       if (strncmp(line, "processor", 9) == 0) this->numProcessors++;
   }
   fclose(file);
-
 }
 
 dlio::OdomNode::~OdomNode() {}
 
-void dlio::OdomNode::getParams() {
-
-  // Version
-  ros::param::param<std::string>("~dlio/version", this->version_, "0.0.0");
-
-  // Frames
-  ros::param::param<std::string>("~dlio/frames/odom", this->odom_frame, "odom");
-  ros::param::param<std::string>("~dlio/frames/baselink", this->baselink_frame, "base_link");
-  ros::param::param<std::string>("~dlio/frames/lidar", this->lidar_frame, "lidar");
-  ros::param::param<std::string>("~dlio/frames/imu", this->imu_frame, "imu");
+void dlio::OdomNode::getParams(const Params &params) {
+  this->verbose = params.verbose;
 
-  // Get Node NS and Remove Leading Character
-  std::string ns = ros::this_node::getNamespace();
-  ns.erase(0,1);
+  this->deskew_ = params.deskew;
+  this->gravity_ = params.gravity;
+  this->time_offset_ = params.time_offset;
 
-  // Concatenate Frame Name Strings
-  this->odom_frame = ns + "/" + this->odom_frame;
-  this->baselink_frame = ns + "/" + this->baselink_frame;
-  this->lidar_frame = ns + "/" + this->lidar_frame;
-  this->imu_frame = ns + "/" + this->imu_frame;
+  this->keyframe_thresh_dist_ = params.keyframe_thresh_dist;
+  this->keyframe_thresh_rot_ = params.keyframe_thresh_rot;
 
-  // Deskew FLag
-  ros::param::param<bool>("~dlio/pointcloud/deskew", this->deskew_, true);
+  this->submap_knn_ = params.submap_knn;
+  this->submap_kcv_ = params.submap_kcv;
+  this->submap_kcc_ = params.submap_kcc;
 
-  // Gravity
-  ros::param::param<double>("~dlio/odom/gravity", this->gravity_, 9.80665);
+  this->densemap_filtered_ = params.densemap_filtered;
 
-  // Compute time offset between lidar and imu
-  ros::param::param<bool>("~dlio/odom/computeTimeOffset", this->time_offset_, false);
+  this->wait_until_move_ = params.wait_until_move;
 
-  // Keyframe Threshold
-  ros::param::param<double>("~dlio/odom/keyframe/threshD", this->keyframe_thresh_dist_, 0.1);
-  ros::param::param<double>("~dlio/odom/keyframe/threshR", this->keyframe_thresh_rot_, 1.0);
+  this->crop_size_ = params.crop_size;
 
-  // Submap
-  ros::param::param<int>("~dlio/odom/submap/keyframe/knn", this->submap_knn_, 10);
-  ros::param::param<int>("~dlio/odom/submap/keyframe/kcv", this->submap_kcv_, 10);
-  ros::param::param<int>("~dlio/odom/submap/keyframe/kcc", this->submap_kcc_, 10);
-
-  // Dense map resolution
-  ros::param::param<bool>("~dlio/map/dense/filtered", this->densemap_filtered_, true);
-
-  // Wait until movement to publish map
-  ros::param::param<bool>("~dlio/map/waitUntilMove", this->wait_until_move_, false);
-
-  // Crop Box Filter
-  ros::param::param<double>("~dlio/odom/preprocessing/cropBoxFilter/size", this->crop_size_, 1.0);
-
-  // Voxel Grid Filter
-  ros::param::param<bool>("~dlio/pointcloud/voxelize", this->vf_use_, true);
-  ros::param::param<double>("~dlio/odom/preprocessing/voxelFilter/res", this->vf_res_, 0.05);
+  this->vf_use_ = params.vf_use;
+  this->vf_res_ = params.vf_res;
 
-  // Adaptive Parameters
-  ros::param::param<bool>("~dlio/adaptive", this->adaptive_params_, true);
+  this->adaptive_params_ = params.adaptive_params;
 
-  // Extrinsics
-  std::vector<float> t_default{0., 0., 0.};
-  std::vector<float> R_default{1., 0., 0., 0., 1., 0., 0., 0., 1.};
-
-  // center of gravity to imu
-  std::vector<float> baselink2imu_t, baselink2imu_R;
-  ros::param::param<std::vector<float>>("~dlio/extrinsics/baselink2imu/t", baselink2imu_t, t_default);
-  ros::param::param<std::vector<float>>("~dlio/extrinsics/baselink2imu/R", baselink2imu_R, R_default);
-  this->extrinsics.baselink2imu.t =
-    Eigen::Vector3f(baselink2imu_t[0], baselink2imu_t[1], baselink2imu_t[2]);
-  this->extrinsics.baselink2imu.R =
-    Eigen::Map<const Eigen::Matrix<float, -1, -1, Eigen::RowMajor>>(baselink2imu_R.data(), 3, 3);
+  this->extrinsics.baselink2imu.t = params.extrinsics_baselink2imu_t;
+  this->extrinsics.baselink2imu.R = params.extrinsics_baselink2imu_R;
+  this->extrinsics.baselink2lidar.t = params.extrinsics_baselink2lidar_t;
+  this->extrinsics.baselink2lidar.R = params.extrinsics_baselink2lidar_R;
 
+  // Build 4x4 transformation matrices
   this->extrinsics.baselink2imu_T = Eigen::Matrix4f::Identity();
-  this->extrinsics.baselink2imu_T.block(0, 3, 3, 1) = this->extrinsics.baselink2imu.t;
-  this->extrinsics.baselink2imu_T.block(0, 0, 3, 3) = this->extrinsics.baselink2imu.R;
-
-  // center of gravity to lidar
-  std::vector<float> baselink2lidar_t, baselink2lidar_R;
-  ros::param::param<std::vector<float>>("~dlio/extrinsics/baselink2lidar/t", baselink2lidar_t, t_default);
-  ros::param::param<std::vector<float>>("~dlio/extrinsics/baselink2lidar/R", baselink2lidar_R, R_default);
-
-  this->extrinsics.baselink2lidar.t =
-    Eigen::Vector3f(baselink2lidar_t[0], baselink2lidar_t[1], baselink2lidar_t[2]);
-  this->extrinsics.baselink2lidar.R =
-    Eigen::Map<const Eigen::Matrix<float, -1, -1, Eigen::RowMajor>>(baselink2lidar_R.data(), 3, 3);
+  this->extrinsics.baselink2imu_T.block(0, 3, 3, 1) =
+      this->extrinsics.baselink2imu.t;
+  this->extrinsics.baselink2imu_T.block(0, 0, 3, 3) =
+      this->extrinsics.baselink2imu.R;
 
   this->extrinsics.baselink2lidar_T = Eigen::Matrix4f::Identity();
-  this->extrinsics.baselink2lidar_T.block(0, 3, 3, 1) = this->extrinsics.baselink2lidar.t;
-  this->extrinsics.baselink2lidar_T.block(0, 0, 3, 3) = this->extrinsics.baselink2lidar.R;
-
-  // IMU
-  ros::param::param<bool>("~dlio/odom/imu/calibration/accel", this->calibrate_accel_, true);
-  ros::param::param<bool>("~dlio/odom/imu/calibration/gyro", this->calibrate_gyro_, true);
-  ros::param::param<double>("~dlio/odom/imu/calibration/time", this->imu_calib_time_, 3.0);
-  ros::param::param<int>("~dlio/odom/imu/bufferSize", this->imu_buffer_size_, 2000);
-
-  std::vector<float> accel_default{0., 0., 0.}; std::vector<float> prior_accel_bias;
-  std::vector<float> gyro_default{0., 0., 0.}; std::vector<float> prior_gyro_bias;
-
-  ros::param::param<bool>("~dlio/odom/imu/approximateGravity", this->gravity_align_, true);
-  ros::param::param<bool>("~dlio/imu/calibration", this->imu_calibrate_, true);
-  ros::param::param<std::vector<float>>("~dlio/imu/intrinsics/accel/bias", prior_accel_bias, accel_default);
-  ros::param::param<std::vector<float>>("~dlio/imu/intrinsics/gyro/bias", prior_gyro_bias, gyro_default);
-
-  // scale-misalignment matrix
-  std::vector<float> imu_sm_default{1., 0., 0., 0., 1., 0., 0., 0., 1.};
-  std::vector<float> imu_sm;
-
-  ros::param::param<std::vector<float>>("~dlio/imu/intrinsics/accel/sm", imu_sm, imu_sm_default);
-
-  if (!this->imu_calibrate_) {
-    this->state.b.accel[0] = prior_accel_bias[0];
-    this->state.b.accel[1] = prior_accel_bias[1];
-    this->state.b.accel[2] = prior_accel_bias[2];
-    this->state.b.gyro[0] = prior_gyro_bias[0];
-    this->state.b.gyro[1] = prior_gyro_bias[1];
-    this->state.b.gyro[2] = prior_gyro_bias[2];
-    this->imu_accel_sm_ = Eigen::Map<const Eigen::Matrix<float, -1, -1, Eigen::RowMajor>>(imu_sm.data(), 3, 3);
-  } else {
-    this->state.b.accel = Eigen::Vector3f(0., 0., 0.);
-    this->state.b.gyro = Eigen::Vector3f(0., 0., 0.);
-    this->imu_accel_sm_ = Eigen::Matrix3f::Identity();
-  }
-
-  // GICP
-  ros::param::param<int>("~dlio/odom/gicp/minNumPoints", this->gicp_min_num_points_, 100);
-  ros::param::param<int>("~dlio/odom/gicp/kCorrespondences", this->gicp_k_correspondences_, 20);
-  ros::param::param<double>("~dlio/odom/gicp/maxCorrespondenceDistance", this->gicp_max_corr_dist_,
-      std::sqrt(std::numeric_limits<double>::max()));
-  ros::param::param<int>("~dlio/odom/gicp/maxIterations", this->gicp_max_iter_, 64);
-  ros::param::param<double>("~dlio/odom/gicp/transformationEpsilon", this->gicp_transformation_ep_, 0.0005);
-  ros::param::param<double>("~dlio/odom/gicp/rotationEpsilon", this->gicp_rotation_ep_, 0.0005);
-  ros::param::param<double>("~dlio/odom/gicp/initLambdaFactor", this->gicp_init_lambda_factor_, 1e-9);
-
-  // Geometric Observer
-  ros::param::param<double>("~dlio/odom/geo/Kp", this->geo_Kp_, 1.0);
-  ros::param::param<double>("~dlio/odom/geo/Kv", this->geo_Kv_, 1.0);
-  ros::param::param<double>("~dlio/odom/geo/Kq", this->geo_Kq_, 1.0);
-  ros::param::param<double>("~dlio/odom/geo/Kab", this->geo_Kab_, 1.0);
-  ros::param::param<double>("~dlio/odom/geo/Kgb", this->geo_Kgb_, 1.0);
-  ros::param::param<double>("~dlio/odom/geo/abias_max", this->geo_abias_max_, 1.0);
-  ros::param::param<double>("~dlio/odom/geo/gbias_max", this->geo_gbias_max_, 1.0);
-
-  ros::param::param<bool>("~dlio/verbose", this->verbose, true);
+  this->extrinsics.baselink2lidar_T.block(0, 3, 3, 1) =
+      this->extrinsics.baselink2lidar.t;
+  this->extrinsics.baselink2lidar_T.block(0, 0, 3, 3) =
+      this->extrinsics.baselink2lidar.R;
+
+  this->calibrate_gyro_ = params.calibrate_gyro;
+  this->calibrate_accel_ = params.calibrate_accel;
+  this->imu_calib_time_ = params.imu_calib_time;
+  this->imu_buffer_size_ = params.imu_buffer_size;
+
+  this->gravity_align_ = params.gravity_align;
+  this->imu_calibrate_ = params.imu_calibrate;
+
+  // Initialize IMU accelerometer scale-misalignment matrix
+  this->state.b.accel = Eigen::Vector3f::Zero();
+  this->state.b.gyro = Eigen::Vector3f::Zero();
+  this->imu_accel_sm_ = Eigen::Matrix3f::Identity();
+
+  this->gicp_min_num_points_ = params.gicp_min_num_points;
+  this->gicp_k_correspondences_ = params.gicp_k_correspondences;
+  this->gicp_max_corr_dist_ = params.gicp_max_corr_dist;
+  this->gicp_max_iter_ = params.gicp_max_iter;
+  this->gicp_transformation_ep_ = params.gicp_transformation_ep;
+  this->gicp_rotation_ep_ = params.gicp_rotation_ep;
+  this->gicp_init_lambda_factor_ = params.gicp_init_lambda_factor;
+
+  this->geo_Kp_ = params.geo_Kp;
+  this->geo_Kv_ = params.geo_Kv;
+  this->geo_Kq_ = params.geo_Kq;
+  this->geo_Kab_ = params.geo_Kab;
+  this->geo_Kgb_ = params.geo_Kgb;
+  this->geo_abias_max_ = params.geo_abias_max;
+  this->geo_gbias_max_ = params.geo_gbias_max;
 }
 
 void dlio::OdomNode::start() {
@@ -318,185 +233,12 @@ void dlio::OdomNode::start() {
 
 }
 
-void dlio::OdomNode::publishPose(const ros::TimerEvent& e) {
-
-  // nav_msgs::Odometry
-  this->odom_ros.header.stamp = this->imu_stamp;
-  this->odom_ros.header.frame_id = this->odom_frame;
-  this->odom_ros.child_frame_id = this->baselink_frame;
-
-  this->odom_ros.pose.pose.position.x = this->state.p[0];
-  this->odom_ros.pose.pose.position.y = this->state.p[1];
-  this->odom_ros.pose.pose.position.z = this->state.p[2];
-
-  this->odom_ros.pose.pose.orientation.w = this->state.q.w();
-  this->odom_ros.pose.pose.orientation.x = this->state.q.x();
-  this->odom_ros.pose.pose.orientation.y = this->state.q.y();
-  this->odom_ros.pose.pose.orientation.z = this->state.q.z();
-
-  this->odom_ros.twist.twist.linear.x = this->state.v.lin.w[0];
-  this->odom_ros.twist.twist.linear.y = this->state.v.lin.w[1];
-  this->odom_ros.twist.twist.linear.z = this->state.v.lin.w[2];
-
-  this->odom_ros.twist.twist.angular.x = this->state.v.ang.b[0];
-  this->odom_ros.twist.twist.angular.y = this->state.v.ang.b[1];
-  this->odom_ros.twist.twist.angular.z = this->state.v.ang.b[2];
-
-  this->odom_pub.publish(this->odom_ros);
-
-  // geometry_msgs::PoseStamped
-  this->pose_ros.header.stamp = this->imu_stamp;
-  this->pose_ros.header.frame_id = this->odom_frame;
-
-  this->pose_ros.pose.position.x = this->state.p[0];
-  this->pose_ros.pose.position.y = this->state.p[1];
-  this->pose_ros.pose.position.z = this->state.p[2];
-
-  this->pose_ros.pose.orientation.w = this->state.q.w();
-  this->pose_ros.pose.orientation.x = this->state.q.x();
-  this->pose_ros.pose.orientation.y = this->state.q.y();
-  this->pose_ros.pose.orientation.z = this->state.q.z();
-
-  this->pose_pub.publish(this->pose_ros);
-
-}
-
-void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published_cloud, Eigen::Matrix4f T_cloud) {
-  this->publishCloud(published_cloud, T_cloud);
-
-  // nav_msgs::Path
-  this->path_ros.header.stamp = this->imu_stamp;
-  this->path_ros.header.frame_id = this->odom_frame;
-
-  geometry_msgs::PoseStamped p;
-  p.header.stamp = this->imu_stamp;
-  p.header.frame_id = this->odom_frame;
-  p.pose.position.x = this->state.p[0];
-  p.pose.position.y = this->state.p[1];
-  p.pose.position.z = this->state.p[2];
-  p.pose.orientation.w = this->state.q.w();
-  p.pose.orientation.x = this->state.q.x();
-  p.pose.orientation.y = this->state.q.y();
-  p.pose.orientation.z = this->state.q.z();
-
-  this->path_ros.poses.push_back(p);
-  this->path_pub.publish(this->path_ros);
-
-  // transform: odom to baselink
-  static tf2_ros::TransformBroadcaster br;
-  geometry_msgs::TransformStamped transformStamped;
-
-  transformStamped.header.stamp = this->imu_stamp;
-  transformStamped.header.frame_id = this->odom_frame;
-  transformStamped.child_frame_id = this->baselink_frame;
-
-  transformStamped.transform.translation.x = this->state.p[0];
-  transformStamped.transform.translation.y = this->state.p[1];
-  transformStamped.transform.translation.z = this->state.p[2];
-
-  transformStamped.transform.rotation.w = this->state.q.w();
-  transformStamped.transform.rotation.x = this->state.q.x();
-  transformStamped.transform.rotation.y = this->state.q.y();
-  transformStamped.transform.rotation.z = this->state.q.z();
-
-  br.sendTransform(transformStamped);
-
-  // transform: baselink to imu
-  transformStamped.header.stamp = this->imu_stamp;
-  transformStamped.header.frame_id = this->baselink_frame;
-  transformStamped.child_frame_id = this->imu_frame;
-
-  transformStamped.transform.translation.x = this->extrinsics.baselink2imu.t[0];
-  transformStamped.transform.translation.y = this->extrinsics.baselink2imu.t[1];
-  transformStamped.transform.translation.z = this->extrinsics.baselink2imu.t[2];
-
-  Eigen::Quaternionf q(this->extrinsics.baselink2imu.R);
-  transformStamped.transform.rotation.w = q.w();
-  transformStamped.transform.rotation.x = q.x();
-  transformStamped.transform.rotation.y = q.y();
-  transformStamped.transform.rotation.z = q.z();
-
-  br.sendTransform(transformStamped);
-
-  // transform: baselink to lidar
-  transformStamped.header.stamp = this->imu_stamp;
-  transformStamped.header.frame_id = this->baselink_frame;
-  transformStamped.child_frame_id = this->lidar_frame;
-
-  transformStamped.transform.translation.x = this->extrinsics.baselink2lidar.t[0];
-  transformStamped.transform.translation.y = this->extrinsics.baselink2lidar.t[1];
-  transformStamped.transform.translation.z = this->extrinsics.baselink2lidar.t[2];
+void dlio::OdomNode::getScanFromROS(
+    const pcl::PointCloud<PointType>::ConstPtr &pc, double stamp) {
 
-  Eigen::Quaternionf qq(this->extrinsics.baselink2lidar.R);
-  transformStamped.transform.rotation.w = qq.w();
-  transformStamped.transform.rotation.x = qq.x();
-  transformStamped.transform.rotation.y = qq.y();
-  transformStamped.transform.rotation.z = qq.z();
-
-  br.sendTransform(transformStamped);
-
-}
-
-void dlio::OdomNode::publishCloud(pcl::PointCloud<PointType>::ConstPtr published_cloud, Eigen::Matrix4f T_cloud) {
-
-  if (this->wait_until_move_) {
-    if (this->length_traversed < 0.1) { return; }
-  }
-
-  pcl::PointCloud<PointType>::Ptr deskewed_scan_t_ (boost::make_shared<pcl::PointCloud<PointType>>());
-
-  pcl::transformPointCloud (*published_cloud, *deskewed_scan_t_, T_cloud);
-
-  // published deskewed cloud
-  sensor_msgs::PointCloud2 deskewed_ros;
-  pcl::toROSMsg(*deskewed_scan_t_, deskewed_ros);
-  deskewed_ros.header.stamp = this->scan_header_stamp;
-  deskewed_ros.header.frame_id = this->odom_frame;
-  this->deskewed_pub.publish(deskewed_ros);
-
-}
-
-void dlio::OdomNode::publishKeyframe(std::pair<std::pair<Eigen::Vector3f, Eigen::Quaternionf>, pcl::PointCloud<PointType>::ConstPtr> kf, ros::Time timestamp) {
-
-  // Push back
-  geometry_msgs::Pose p;
-  p.position.x = kf.first.first[0];
-  p.position.y = kf.first.first[1];
-  p.position.z = kf.first.first[2];
-  p.orientation.w = kf.first.second.w();
-  p.orientation.x = kf.first.second.x();
-  p.orientation.y = kf.first.second.y();
-  p.orientation.z = kf.first.second.z();
-  this->kf_pose_ros.poses.push_back(p);
-
-  // Publish
-  this->kf_pose_ros.header.stamp = timestamp;
-  this->kf_pose_ros.header.frame_id = this->odom_frame;
-  this->kf_pose_pub.publish(this->kf_pose_ros);
-
-  // publish keyframe scan for map
-  if (this->vf_use_) {
-    if (kf.second->points.size() == kf.second->width * kf.second->height) {
-      sensor_msgs::PointCloud2 keyframe_cloud_ros;
-      pcl::toROSMsg(*kf.second, keyframe_cloud_ros);
-      keyframe_cloud_ros.header.stamp = timestamp;
-      keyframe_cloud_ros.header.frame_id = this->odom_frame;
-      this->kf_cloud_pub.publish(keyframe_cloud_ros);
-    }
-  } else {
-    sensor_msgs::PointCloud2 keyframe_cloud_ros;
-    pcl::toROSMsg(*kf.second, keyframe_cloud_ros);
-    keyframe_cloud_ros.header.stamp = timestamp;
-    keyframe_cloud_ros.header.frame_id = this->odom_frame;
-    this->kf_cloud_pub.publish(keyframe_cloud_ros);
-  }
-
-}
-
-void dlio::OdomNode::getScanFromROS(const sensor_msgs::PointCloud2ConstPtr& pc) {
-
-  pcl::PointCloud<PointType>::Ptr original_scan_ (boost::make_shared<pcl::PointCloud<PointType>>());
-  pcl::fromROSMsg(*pc, *original_scan_);
+  pcl::PointCloud<PointType>::Ptr original_scan_(
+      std::make_shared<pcl::PointCloud<PointType>>());
+  *original_scan_ = *pc;
 
   // Remove NaNs
   std::vector<int> idx;
@@ -508,28 +250,14 @@ void dlio::OdomNode::getScanFromROS(const sensor_msgs::PointCloud2ConstPtr& pc)
   this->crop.filter(*original_scan_);
 
   // automatically detect sensor type
-  this->sensor = dlio::SensorType::UNKNOWN;
-  for (auto &field : pc->fields) {
-    if (field.name == "t") {
-      this->sensor = dlio::SensorType::OUSTER;
-      break;
-    } else if (field.name == "time") {
-      this->sensor = dlio::SensorType::VELODYNE;
-      break;
-    } else if (field.name == "timestamp" && original_scan_->points[0].timestamp < 1e14) {
-      this->sensor = dlio::SensorType::HESAI;
-      break;
-    } else if (field.name == "timestamp" && original_scan_->points[0].timestamp > 1e14) {
-      this->sensor = dlio::SensorType::LIVOX;
-      break;
-    }
-  }
-
+  // NOTE: Ouster used here as it utilizes time since start of scan in
+  // nanoseconds like evalio
+  this->sensor = dlio::SensorType::OUSTER;
   if (this->sensor == dlio::SensorType::UNKNOWN) {
     this->deskew_ = false;
   }
 
-  this->scan_header_stamp = pc->header.stamp;
+  this->scan_header_stamp = stamp;
   this->original_scan = original_scan_;
 
 }
@@ -547,7 +275,7 @@ void dlio::OdomNode::preprocessPoints() {
 
   } else {
 
-    this->scan_stamp = this->scan_header_stamp.toSec();
+    this->scan_stamp = this->scan_header_stamp;
 
     // don't process scans until IMU data is present
     if (!this->first_valid_scan) {
@@ -574,7 +302,7 @@ void dlio::OdomNode::preprocessPoints() {
 
     }
 
-    pcl::PointCloud<PointType>::Ptr deskewed_scan_ (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr deskewed_scan_ (std::make_shared<pcl::PointCloud<PointType>>());
     pcl::transformPointCloud (*this->original_scan, *deskewed_scan_,
                               this->T_prior * this->extrinsics.baselink2lidar_T);
     this->deskewed_scan = deskewed_scan_;
@@ -584,7 +312,7 @@ void dlio::OdomNode::preprocessPoints() {
   // Voxel Grid Filter
   if (this->vf_use_) {
     pcl::PointCloud<PointType>::Ptr current_scan_
-      (boost::make_shared<pcl::PointCloud<PointType>>(*this->deskewed_scan));
+      (std::make_shared<pcl::PointCloud<PointType>>(*this->deskewed_scan));
     this->voxel.setInputCloud(current_scan_);
     this->voxel.filter(*current_scan_);
     this->current_scan = current_scan_;
@@ -596,11 +324,11 @@ void dlio::OdomNode::preprocessPoints() {
 
 void dlio::OdomNode::deskewPointcloud() {
 
-  pcl::PointCloud<PointType>::Ptr deskewed_scan_ (boost::make_shared<pcl::PointCloud<PointType>>());
+  pcl::PointCloud<PointType>::Ptr deskewed_scan_ (std::make_shared<pcl::PointCloud<PointType>>());
   deskewed_scan_->points.resize(this->original_scan->points.size());
 
   // individual point timestamps should be relative to this time
-  double sweep_ref_time = this->scan_header_stamp.toSec();
+  double sweep_ref_time = this->scan_header_stamp;
 
   // sort points by timestamp and build list of timestamps
   std::function<bool(const PointType&, const PointType&)> point_time_cmp;
@@ -700,8 +428,12 @@ void dlio::OdomNode::deskewPointcloud() {
   // if there are no frames between the start and end of the sweep
   // that probably means that there's a sync issue
   if (frames.size() != timestamps.size()) {
-    ROS_FATAL("Bad time sync between LiDAR and IMU!");
-
+    if(verbose) {
+      printf("FATAL: ");
+      printf("Bad time sync between LiDAR and IMU!");
+      printf("\n");
+    }
+    
     this->T_prior = this->T;
     pcl::transformPointCloud (*deskewed_scan_, *deskewed_scan_, this->T_prior * this->extrinsics.baselink2lidar_T);
     this->deskewed_scan = deskewed_scan_;
@@ -755,20 +487,25 @@ void dlio::OdomNode::initializeDLIO() {
   }
 
   this->dlio_initialized = true;
-  std::cout << std::endl << " DLIO initialized!" << std::endl;
+  if (verbose) {
+    std::cout << std::endl << " DLIO initialized!" << std::endl;
+  }
 
 }
 
-void dlio::OdomNode::callbackPointCloud(const sensor_msgs::PointCloud2ConstPtr& pc) {
+void dlio::OdomNode::callbackPointCloud(
+    const pcl::PointCloud<PointType>::ConstPtr &pc, double stamp) {
 
   std::unique_lock<decltype(this->main_loop_running_mutex)> lock(main_loop_running_mutex);
   this->main_loop_running = true;
   lock.unlock();
 
-  double then = ros::Time::now().toSec();
+  double then = std::chrono::duration_cast<std::chrono::duration<double>>(
+                    std::chrono::system_clock::now().time_since_epoch())
+                    .count();
 
   if (this->first_scan_stamp == 0.) {
-    this->first_scan_stamp = pc->header.stamp.toSec();
+    this->first_scan_stamp = stamp;
   }
 
   // DLIO Initialization procedures (IMU calib, gravity align)
@@ -777,7 +514,7 @@ void dlio::OdomNode::callbackPointCloud(const sensor_msgs::PointCloud2ConstPtr&
   }
 
   // Convert incoming scan into DLIO format
-  this->getScanFromROS(pc);
+  this->getScanFromROS(pc, stamp);
 
   // Preprocess points
   this->preprocessPoints();
@@ -787,13 +524,16 @@ void dlio::OdomNode::callbackPointCloud(const sensor_msgs::PointCloud2ConstPtr&
   }
 
   if (this->current_scan->points.size() <= this->gicp_min_num_points_) {
-    ROS_FATAL("Low number of points in the cloud!");
-    return;
+    if (verbose) {
+      printf("FATAL: ");
+      printf("Low number of points in the cloud!");
+      printf("\n");
+      return;
+    }
   }
 
   // Compute Metrics
-  this->metrics_thread = std::thread( &dlio::OdomNode::computeMetrics, this );
-  this->metrics_thread.detach();
+  this->computeMetrics();
 
   // Set Adaptive Parameters
   if (this->adaptive_params_) {
@@ -839,50 +579,43 @@ void dlio::OdomNode::callbackPointCloud(const sensor_msgs::PointCloud2ConstPtr&
   this->prev_scan_stamp = this->scan_stamp;
   this->elapsed_time = this->scan_stamp - this->first_scan_stamp;
 
-  // Publish stuff to ROS
-  pcl::PointCloud<PointType>::ConstPtr published_cloud;
-  if (this->densemap_filtered_) {
-    published_cloud = this->current_scan;
-  } else {
-    published_cloud = this->deskewed_scan;
-  }
-  this->publish_thread = std::thread( &dlio::OdomNode::publishToROS, this, published_cloud, this->T_corr );
-  this->publish_thread.detach();
-
   // Update some statistics
-  this->comp_times.push_back(ros::Time::now().toSec() - then);
+  this->comp_times.push_back(
+      std::chrono::duration_cast<std::chrono::duration<double>>(
+          std::chrono::system_clock::now().time_since_epoch())
+          .count() -
+      then);
   this->gicp_hasConverged = this->gicp.hasConverged();
 
   // Debug statements and publish custom DLIO message
   if (this->verbose) {
-    this->debug_thread = std::thread( &dlio::OdomNode::debug, this );
-    this->debug_thread.detach();
+    this->debug();
   }
   
   this->geo.first_opt_done = true;
 }
 
-void dlio::OdomNode::callbackImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
+void dlio::OdomNode::callbackImu(const ImuMeas &imu_raw) {
 
   this->first_imu_received = true;
 
-  sensor_msgs::Imu::Ptr imu = this->transformImu( imu_raw );
-  this->imu_stamp = imu->header.stamp;
+  ImuMeas imu = this->transformImu(imu_raw);
+  this->imu_stamp = imu.stamp;
 
   Eigen::Vector3f lin_accel;
   Eigen::Vector3f ang_vel;
 
   // Get IMU samples
-  ang_vel[0] = imu->angular_velocity.x;
-  ang_vel[1] = imu->angular_velocity.y;
-  ang_vel[2] = imu->angular_velocity.z;
+  ang_vel[0] = imu.ang_vel[0];
+  ang_vel[1] = imu.ang_vel[1];
+  ang_vel[2] = imu.ang_vel[2];
 
-  lin_accel[0] = imu->linear_acceleration.x;
-  lin_accel[1] = imu->linear_acceleration.y;
-  lin_accel[2] = imu->linear_acceleration.z;
+  lin_accel[0] = imu.lin_accel[0];
+  lin_accel[1] = imu.lin_accel[1];
+  lin_accel[2] = imu.lin_accel[2];
 
   if (this->first_imu_stamp == 0.) {
-    this->first_imu_stamp = imu->header.stamp.toSec();
+    this->first_imu_stamp = imu.stamp;
   }
 
   // IMU calibration procedure - do for three seconds
@@ -893,7 +626,7 @@ void dlio::OdomNode::callbackImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
     static Eigen::Vector3f accel_avg (0., 0., 0.);
     static bool print = true;
 
-    if ((imu->header.stamp.toSec() - this->first_imu_stamp) < this->imu_calib_time_) {
+    if ((imu.stamp - this->first_imu_stamp) < this->imu_calib_time_) {
 
       num_samples++;
 
@@ -905,7 +638,7 @@ void dlio::OdomNode::callbackImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
       accel_avg[1] += lin_accel[1];
       accel_avg[2] += lin_accel[2];
 
-      if(print) {
+      if(verbose) {
         std::cout << std::endl << " Calibrating IMU for " << this->imu_calib_time_ << " seconds... ";
         std::cout.flush();
         print = false;
@@ -913,7 +646,9 @@ void dlio::OdomNode::callbackImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
 
     } else {
 
-      std::cout << "done" << std::endl << std::endl;
+      if(verbose) {
+        std::cout << "done" << std::endl << std::endl;
+      }
 
       gyro_avg /= num_samples;
       accel_avg /= num_samples;
@@ -943,44 +678,47 @@ void dlio::OdomNode::callbackImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
           pitch = remainder(180.0 - pitch, 360.0);
           roll  = remainder(roll + 180.0,  360.0);
         }
-        std::cout << " Estimated initial attitude:" << std::endl;
-        std::cout << "   Roll  [deg]: " << to_string_with_precision(roll, 4) << std::endl;
-        std::cout << "   Pitch [deg]: " << to_string_with_precision(pitch, 4) << std::endl;
-        std::cout << "   Yaw   [deg]: " << to_string_with_precision(yaw, 4) << std::endl;
-        std::cout << std::endl;
+        if (verbose){
+          std::cout << " Estimated initial attitude:" << std::endl;
+          std::cout << "   Roll  [deg]: " << to_string_with_precision(roll, 4) << std::endl;
+          std::cout << "   Pitch [deg]: " << to_string_with_precision(pitch, 4) << std::endl;
+          std::cout << "   Yaw   [deg]: " << to_string_with_precision(yaw, 4) << std::endl;
+          std::cout << std::endl;
+        }
       }
 
       if (this->calibrate_accel_) {
 
         // subtract gravity from avg accel to get bias
         this->state.b.accel = accel_avg - grav_vec;
-
-        std::cout << " Accel biases [xyz]: " << to_string_with_precision(this->state.b.accel[0], 8) << ", "
-                                             << to_string_with_precision(this->state.b.accel[1], 8) << ", "
-                                             << to_string_with_precision(this->state.b.accel[2], 8) << std::endl;
+        if (verbose) {
+          std::cout << " Accel biases [xyz]: " << to_string_with_precision(this->state.b.accel[0], 8) << ", "
+                                               << to_string_with_precision(this->state.b.accel[1], 8) << ", "
+                                               << to_string_with_precision(this->state.b.accel[2], 8) << std::endl;
+        }
       }
 
       if (this->calibrate_gyro_) {
 
         this->state.b.gyro = gyro_avg;
-
-        std::cout << " Gyro biases  [xyz]: " << to_string_with_precision(this->state.b.gyro[0], 8) << ", "
-                                             << to_string_with_precision(this->state.b.gyro[1], 8) << ", "
-                                             << to_string_with_precision(this->state.b.gyro[2], 8) << std::endl;
+        if(verbose) {
+          std::cout << " Gyro biases  [xyz]: " << to_string_with_precision(this->state.b.gyro[0], 8) << ", "
+                                               << to_string_with_precision(this->state.b.gyro[1], 8) << ", "
+                                               << to_string_with_precision(this->state.b.gyro[2], 8) << std::endl;
+        }
       }
 
       this->imu_calibrated = true;
-
     }
 
   } else {
 
-    double dt = imu->header.stamp.toSec() - this->prev_imu_stamp;
+    double dt = imu.stamp - this->prev_imu_stamp;
     if (dt == 0) { dt = 1.0/200.0; }
     this->imu_rates.push_back( 1./dt );
 
     // Apply the calibrated bias to the new IMU measurements
-    this->imu_meas.stamp = imu->header.stamp.toSec();
+    this->imu_meas.stamp = imu.stamp;
     this->imu_meas.dt = dt;
     this->prev_imu_stamp = this->imu_meas.stamp;
 
@@ -990,7 +728,8 @@ void dlio::OdomNode::callbackImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
     this->imu_meas.lin_accel = lin_accel_corrected;
     this->imu_meas.ang_vel = ang_vel_corrected;
 
-    // Store calibrated IMU measurements into imu buffer for manual integration later.
+    // Store calibrated IMU measurements into imu buffer for manual integration
+    // later.
     this->mtx_imu.lock();
     this->imu_buffer.push_front(this->imu_meas);
     this->mtx_imu.unlock();
@@ -1027,7 +766,7 @@ void dlio::OdomNode::getNextPose() {
   }
 
   // Align with current submap with global IMU transformation as initial guess
-  pcl::PointCloud<PointType>::Ptr aligned (boost::make_shared<pcl::PointCloud<PointType>>());
+  pcl::PointCloud<PointType>::Ptr aligned (std::make_shared<pcl::PointCloud<PointType>>());
   this->gicp.align(*aligned);
 
   // Get final transformation in global frame
@@ -1046,11 +785,12 @@ void dlio::OdomNode::getNextPose() {
 bool dlio::OdomNode::imuMeasFromTimeRange(double start_time, double end_time,
                                           boost::circular_buffer<ImuMeas>::reverse_iterator& begin_imu_it,
                                           boost::circular_buffer<ImuMeas>::reverse_iterator& end_imu_it) {
-
+  // Evalio run deterministically, nothing will ever appear here!
   if (this->imu_buffer.empty() || this->imu_buffer.front().stamp < end_time) {
-    // Wait for the latest IMU data
-    std::unique_lock<decltype(this->mtx_imu)> lock(this->mtx_imu);
-    this->cv_imu_stamp.wait(lock, [this, &end_time]{ return this->imu_buffer.front().stamp >= end_time; });
+    // // Wait for the latest IMU data
+    // std::unique_lock<decltype(this->mtx_imu)> lock(this->mtx_imu);
+    // this->cv_imu_stamp.wait(lock, [this, &end_time]{ return this->imu_buffer.front().stamp >= end_time; });
+    return false;
   }
 
   auto imu_it = this->imu_buffer.begin();
@@ -1268,7 +1008,6 @@ void dlio::OdomNode::propagateGICP() {
   double norm = sqrt(q.w()*q.w() + q.x()*q.x() + q.y()*q.y() + q.z()*q.z());
   q.w() /= norm; q.x() /= norm; q.y() /= norm; q.z() /= norm;
   this->lidarPose.q = q;
-
 }
 
 void dlio::OdomNode::propagateState() {
@@ -1369,36 +1108,36 @@ void dlio::OdomNode::updateState() {
 
 }
 
-sensor_msgs::Imu::Ptr dlio::OdomNode::transformImu(const sensor_msgs::Imu::ConstPtr& imu_raw) {
+dlio::OdomNode::ImuMeas dlio::OdomNode::transformImu(const ImuMeas &imu_raw) {
 
-  sensor_msgs::Imu::Ptr imu (new sensor_msgs::Imu);
+  ImuMeas imu;
 
   // Copy header
-  imu->header = imu_raw->header;
+  imu.stamp = imu_raw.stamp;
 
-  static double prev_stamp = imu->header.stamp.toSec();
-  double dt = imu->header.stamp.toSec() - prev_stamp;
-  prev_stamp = imu->header.stamp.toSec();
+  static double prev_stamp = imu.stamp;
+  double dt = imu.stamp - prev_stamp;
+  prev_stamp = imu.stamp;
   
   if (dt == 0) { dt = 1.0/200.0; }
 
-  // Transform angular velocity (will be the same on a rigid body, so just rotate to ROS convention)
-  Eigen::Vector3f ang_vel(imu_raw->angular_velocity.x,
-                          imu_raw->angular_velocity.y,
-                          imu_raw->angular_velocity.z);
+  // Transform angular velocity (will be the same on a rigid body, so just
+  // rotate to ROS convention)
+  Eigen::Vector3f ang_vel(imu_raw.ang_vel[0], imu_raw.ang_vel[1],
+                          imu_raw.ang_vel[2]);
 
   Eigen::Vector3f ang_vel_cg = this->extrinsics.baselink2imu.R * ang_vel;
 
-  imu->angular_velocity.x = ang_vel_cg[0];
-  imu->angular_velocity.y = ang_vel_cg[1];
-  imu->angular_velocity.z = ang_vel_cg[2];
+  imu.ang_vel[0] = ang_vel_cg[0];
+  imu.ang_vel[1] = ang_vel_cg[1];
+  imu.ang_vel[2] = ang_vel_cg[2];
 
   static Eigen::Vector3f ang_vel_cg_prev = ang_vel_cg;
 
-  // Transform linear acceleration (need to account for component due to translational difference)
-  Eigen::Vector3f lin_accel(imu_raw->linear_acceleration.x,
-                            imu_raw->linear_acceleration.y,
-                            imu_raw->linear_acceleration.z);
+  // Transform linear acceleration (need to account for component due to
+  // translational difference)
+  Eigen::Vector3f lin_accel(imu_raw.lin_accel[0], imu_raw.lin_accel[1],
+                            imu_raw.lin_accel[2]);
 
   Eigen::Vector3f lin_accel_cg = this->extrinsics.baselink2imu.R * lin_accel;
 
@@ -1408,9 +1147,9 @@ sensor_msgs::Imu::Ptr dlio::OdomNode::transformImu(const sensor_msgs::Imu::Const
 
   ang_vel_cg_prev = ang_vel_cg;
 
-  imu->linear_acceleration.x = lin_accel_cg[0];
-  imu->linear_acceleration.y = lin_accel_cg[1];
-  imu->linear_acceleration.z = lin_accel_cg[2];
+  imu.lin_accel[0] = lin_accel_cg[0];
+  imu.lin_accel[1] = lin_accel_cg[1];
+  imu.lin_accel[2] = lin_accel_cg[2];
 
   return imu;
 
@@ -1471,7 +1210,7 @@ void dlio::OdomNode::computeConvexHull() {
 
   // create a pointcloud with points at keyframes
   pcl::PointCloud<PointType>::Ptr cloud =
-    pcl::PointCloud<PointType>::Ptr (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr (std::make_shared<pcl::PointCloud<PointType>>());
 
   std::unique_lock<decltype(this->keyframes_mutex)> lock(this->keyframes_mutex);
   for (int i = 0; i < this->num_processed_keyframes; i++) {
@@ -1488,10 +1227,10 @@ void dlio::OdomNode::computeConvexHull() {
 
   // get the indices of the keyframes on the convex hull
   pcl::PointCloud<PointType>::Ptr convex_points =
-    pcl::PointCloud<PointType>::Ptr (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr (std::make_shared<pcl::PointCloud<PointType>>());
   this->convex_hull.reconstruct(*convex_points);
 
-  pcl::PointIndices::Ptr convex_hull_point_idx = pcl::PointIndices::Ptr (boost::make_shared<pcl::PointIndices>());
+  pcl::PointIndices::Ptr convex_hull_point_idx = pcl::PointIndices::Ptr (std::make_shared<pcl::PointIndices>());
   this->convex_hull.getHullPointIndices(*convex_hull_point_idx);
 
   this->keyframe_convex.clear();
@@ -1510,7 +1249,7 @@ void dlio::OdomNode::computeConcaveHull() {
 
   // create a pointcloud with points at keyframes
   pcl::PointCloud<PointType>::Ptr cloud =
-    pcl::PointCloud<PointType>::Ptr (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr (std::make_shared<pcl::PointCloud<PointType>>());
 
   std::unique_lock<decltype(this->keyframes_mutex)> lock(this->keyframes_mutex);
   for (int i = 0; i < this->num_processed_keyframes; i++) {
@@ -1527,10 +1266,10 @@ void dlio::OdomNode::computeConcaveHull() {
 
   // get the indices of the keyframes on the concave hull
   pcl::PointCloud<PointType>::Ptr concave_points =
-    pcl::PointCloud<PointType>::Ptr (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr (std::make_shared<pcl::PointCloud<PointType>>());
   this->concave_hull.reconstruct(*concave_points);
 
-  pcl::PointIndices::Ptr concave_hull_point_idx = pcl::PointIndices::Ptr (boost::make_shared<pcl::PointIndices>());
+  pcl::PointIndices::Ptr concave_hull_point_idx = pcl::PointIndices::Ptr (std::make_shared<pcl::PointIndices>());
   this->concave_hull.getHullPointIndices(*concave_hull_point_idx);
 
   this->keyframe_concave.clear();
@@ -1739,7 +1478,7 @@ void dlio::OdomNode::buildSubmap(State vehicle_state) {
     this->pauseSubmapBuildIfNeeded();
 
     // reinitialize submap cloud and normals
-    pcl::PointCloud<PointType>::Ptr submap_cloud_ (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr submap_cloud_ (std::make_shared<pcl::PointCloud<PointType>>());
     std::shared_ptr<nano_gicp::CovarianceList> submap_normals_ (std::make_shared<nano_gicp::CovarianceList>());
 
     for (auto k : this->submap_kf_idx_curr) {
@@ -1780,7 +1519,7 @@ void dlio::OdomNode::buildKeyframesAndSubmap(State vehicle_state) {
 
     Eigen::Matrix4d Td = T.cast<double>();
 
-    pcl::PointCloud<PointType>::Ptr transformed_keyframe (boost::make_shared<pcl::PointCloud<PointType>>());
+    pcl::PointCloud<PointType>::Ptr transformed_keyframe (std::make_shared<pcl::PointCloud<PointType>>());
     pcl::transformPointCloud (*raw_keyframe, *transformed_keyframe, T);
 
     std::shared_ptr<nano_gicp::CovarianceList> transformed_covariances (std::make_shared<nano_gicp::CovarianceList>(raw_covariances->size()));
@@ -1792,9 +1531,6 @@ void dlio::OdomNode::buildKeyframesAndSubmap(State vehicle_state) {
     lock.lock();
     this->keyframes[i].second = transformed_keyframe;
     this->keyframe_normals[i] = transformed_covariances;
-
-    this->publish_keyframe_thread = std::thread( &dlio::OdomNode::publishKeyframe, this, this->keyframes[i], this->keyframe_timestamps[i] );
-    this->publish_keyframe_thread.detach();
   }
 
   lock.unlock();
@@ -2018,3 +1754,19 @@ void dlio::OdomNode::debug() {
   std::cout << "+-------------------------------------------------------------------+" << std::endl;
 
 }
+
+// Explicit instantiations of PCL templates for dlio::Point
+// This is needed because PCL doesn't automatically instantiate templates for
+// custom point types
+#include <pcl/kdtree/impl/kdtree_flann.hpp>
+#include <pcl/kdtree/kdtree_flann.h>
+#include <pcl/pcl_base.h>
+#include <pcl/search/impl/kdtree.hpp>
+#include <pcl/search/kdtree.h>
+#include <pcl/search/search.h>
+
+template class pcl::PCLBase<dlio::Point>;
+template class pcl::search::Search<dlio::Point>;
+template class pcl::KdTreeFLANN<dlio::Point, flann::L2_Simple<float>>;
+template class pcl::search::KdTree<
+    dlio::Point, pcl::KdTreeFLANN<dlio::Point, flann::L2_Simple<float>>>;
\ No newline at end of file
diff --git a/src/nano_gicp/lsq_registration.cc b/src/nano_gicp/lsq_registration.cc
index 271cb03..b6c8f07 100644
--- a/src/nano_gicp/lsq_registration.cc
+++ b/src/nano_gicp/lsq_registration.cc
@@ -42,10 +42,8 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *************************************************************************/
 
-#include "dlio/dlio.h"
 #include "nano_gicp/lsq_registration.h"
-
-template class nano_gicp::LsqRegistration<PointType, PointType>;
+#include "dlio/dlio.h"
 
 namespace nano_gicp {
 
@@ -228,4 +226,7 @@ bool LsqRegistration<PointTarget, PointSource>::step_lm(Eigen::Isometry3d& x0, E
   return false;
 }
 
-}  // namespace nano_gicp
+} // namespace nano_gicp
+
+// Explicit template instantiation must come after all member definitions
+template class nano_gicp::LsqRegistration<PointType, PointType>;
diff --git a/src/nano_gicp/nano_gicp.cc b/src/nano_gicp/nano_gicp.cc
index 906e3c3..b9c67a4 100644
--- a/src/nano_gicp/nano_gicp.cc
+++ b/src/nano_gicp/nano_gicp.cc
@@ -42,10 +42,8 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *************************************************************************/
 
-#include "dlio/dlio.h"
 #include "nano_gicp/nano_gicp.h"
-
-template class nano_gicp::NanoGICP<PointType, PointType>;
+#include "dlio/dlio.h"
 
 namespace nano_gicp {
 
@@ -391,4 +389,7 @@ bool NanoGICP<PointSource, PointTarget>::calculate_covariances(
   return true;
 }
 
-}  // namespace nano_gicp
+} // namespace nano_gicp
+
+// Explicit template instantiation must come after all member definitions
+template class nano_gicp::NanoGICP<PointType, PointType>;
