diff --git a/CMakeLists.txt b/CMakeLists.txt
index 84fa788..5d2760b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,6 +19,7 @@ FetchContent_Declare(
   ceres
   GIT_REPOSITORY https://github.com/ceres-solver/ceres-solver.git
   GIT_TAG        2.2.0
+  FIND_PACKAGE_ARGS NO_MODULE # uses find_package first, git if it fails
 )
 FetchContent_MakeAvailable(ceres)
 
@@ -29,6 +30,7 @@ FetchContent_Declare(
   nanoflann
   GIT_REPOSITORY https://github.com/jlblancoc/nanoflann.git
   GIT_TAG        v1.5.5
+  FIND_PACKAGE_ARGS NO_MODULE # uses find_package first, git if it fails
 )
 FetchContent_MakeAvailable(nanoflann)
 
diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index 5f19205..3e4ad5a 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -5,6 +5,7 @@ FetchContent_Declare(
   pybind11
   GIT_REPOSITORY https://github.com/pybind/pybind11.git
   GIT_TAG        v2.11.0
+  FIND_PACKAGE_ARGS NO_MODULE # uses find_package first, git if it fails
 )
 FetchContent_MakeAvailable(pybind11)
 
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index f4712b1..a67bef8 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -5,6 +5,7 @@ FetchContent_Declare(
     googletest
     GIT_REPOSITORY https://github.com/google/googletest.git
     GIT_TAG        origin/main
+    FIND_PACKAGE_ARGS NO_MODULE # uses find_package first, git if it fails
 )
 option(INSTALL_GTEST OFF)
 option(BUILD_GMOCK OFF)
